docs/concepts.md

```diff
@@ -0,0 +1,320 @@
+not directly in their own scope of responsibility. For example, a Request would
+need to listen for the activation of Replicas, so that it could schedule itself
+to switch to the RequestProcessing state.
+
+Instead, as noted above, the responsibility for selecting which Entity to move
+rests with the Source Stock.
+
+#### Relationship to the Systems Dynamics concept of Flows
+
+In System Dynamics there's no "movement" concept and generally Entities are
+represented by a count or summation in a Stock. A Stock in System Dynamics is
+connected to other Stocks via Flows. Like Stocks these are functions of time
+elapsed, but they are derivatives, not integrals: rates of change, rather than
+accumulated amounts of change.
+
+This makes sense in a continuous setting, but does not make sense in a discrete
+setting. In this respect Movements can be thought of as discretized Flows, or
+alternatively as the inverse of a Flow. Where a Flow might expressed "X units
+per second", a Movement will instead look at the seconds until a particular
+Movement is to occur. When aggregated by time, Movements approximate a Flow.
+
+### Models
+
+Models are "the rest" of the code. Typically these own Stocks, wire dependencies
+and potentially maintain other state.
+
+In Skenario the main toplevel Models are the KnativeAutoscaler and the Cluster.
+These establish the various initial Stocks (eg. RequestsBuffered, ReplicasLaunching)
+that are used during the life of the simulation.
+
+Generally speaking, keep logic out of Models. Everything that can be expressed
+as a Movement ought to be; special-case logic should be placed in `Add()` or
+`Remove()`.
+
+### Environment
+
+The Environment is the effective root of the program. Its `Run()` method has the
+core simulation loop and the `AddToSchedule()` method provides the means for
+simulation components to set up future Movements. The Environment also provides a
+number of accessors to key global variables: the simulation's `CurrentMovementTime()`,
+the simulation's `HaltTime()` and a `Context()` that holds a logger object as a value.
+
+The Environment schedules two specialised Movements: the `start_to_running` Movement
+and the `running_to_halted` Movement. These are placed at the boundary points of time.
+That is: only the start Movement may occur at time zero, only the halt Movement may
+occur at the halting time. Other Movements must occur between these two, meaning that
+the permissible time range is expressible as `(0, halt)` or as `[1, halt-1]`.
+
+## How a scenario is executed
+
+### `Run()`
+
+The Environment's `Run()` method implements the "next-event list" concept from
+Discrete Event Simulation. This means that on each iteration, the Environment will
+select the next Movement from a queue and execute it.
+
+Ordering is by the `OccursAt` time of Movements. Internally, the Environment is
+relying on a `MovementPriorityQueue` to maintain orderly records; ultimately this is
+a basic heap.
```

Comment:
Not quite sure what you mean by a "basic heap", but is such a thing ordered and, if so, how?

Memory allocation heaps, which are probably the kinds of heap most familiar to readers, are unordered.